### DISCUSSION TITLE
Passing GitHub secrets to Compose file used from a GitHub Action

### DISCUSSION THREAD

**Discussion by why-not-try-calmer** [community/community#25269] [Actions]
<p>I am trying to get a bunch of secrets defined from my GitHub repository passed to my containerized application (for now the test build). The furthest I have been able to go was by exposing the secrets to a GitHub action responsible for making the test build:</p>
<pre><code class="lang-auto">name: GitHub Test suite

on: 
  pull_request:
    branches: [master]
  push:
    branches: [master]
  workflow_dispatch:

jobs:
  build:
    runs-on: 'ubuntu-latest'
    env:
       TELEGRAM_TOKEN: ${{ secrets.TELEGRAM_TOKEN }}
       BASE_URL: ${{ secrets.BASE_URL }}
       ALERT_CHATID: ${{ secrets.ALERT_CHATID }}
       WEBHOOK_URL: ${{ secrets.WEBHOOK_URL }}
       MONGO_CONN_STRING: ${{ secrets.MONGO_CONN_STRING }}
    steps:
    - uses: actions/checkout@v2

    - name: Set up Docker Buildx
      uses: docker/setup-buildx-action@v1
      
    - name: Build &amp; run tests
      run: |
        docker-compose -f test-build.yml up \
          --abort-on-container-exit
</code></pre>
<p>I tried to fetch these values from the ‘test-build.yml’ file:</p>
<pre><code class="lang-auto">version: "3"
services:
    feedfarer:
        depends_on: [redis]
        restart: "no"
        build: .
        command: stack test
        ports: [80:80]
        environment:
         - TELEGRAM_TOKEN=${TELEGRAM_TOKEN}
         - BASE_URL=${BASE_URL}
         - ALERT_CHATID=${ALERT_CHATID}
         - WEBHOOK_URL=${WEBHOOK_URL}
         - MONGO_CONN_STRING=${MONGO_CONN_STRING}
    redis:
        image: redis:latest
        ports: [6379:6379]
        restart: unless-stopped
        volumes:
          - /data/redis-store

</code></pre>
<p>… but it does not work. The application reports empty strings on all these values.</p>
<p>Am I going about this the wrong way? Any help or advice appreciated.</p>

**Comment by airtower-luna** [comment]
<p>How did you test the workflow? In particular, was it via a PR from a fork? In that case the workflow won’t have access to secrets.</p>
<p>Unlikely to be the cause, but still highly recommended for security: When putting secrets into environment variables do it only for the step that needs them, not the whole job. That minimizes the risk of them getting anywhere you don’t want them.</p>

**Comment by why-not-try-calmer** [comment]
<p>Hello, thanks for your reply and suggestion.</p>
<p><em>How did I test the worflow?</em><br>
I’ve tested it on my project. Everything works fine EXCEPT for the part that is at issue; namely, I cannot access the aforementioned envars from the containerized application.</p>
<p><em>I should restrict envar definitions to the site that actually uses them</em><br>
Okay, I’ve tried replacing ‘test-build’ with:</p>
<pre><code class="lang-auto">
on: 
  pull_request:
    branches: [master]
  push:
    branches: [master]
  workflow_dispatch:

jobs:
  build:
    runs-on: 'ubuntu-latest'
    
    steps:
    - uses: actions/checkout@v2

    - name: Set up Docker Buildx
      uses: docker/setup-buildx-action@v1
      
    - name: Build &amp; run tests
      env:
       TELEGRAM_TOKEN: ${{ secrets.TELEGRAM_TOKEN }}
       BASE_URL: ${{ secrets.BASE_URL }}
       ALERT_CHATID: ${{ secrets.ALERT_CHATID }}
       WEBHOOK_URL: ${{ secrets.WEBHOOK_URL }}
       MONGO_CONN_STRING: ${{ secrets.MONGO_CONN_STRING }}
      run: |
        docker-compose -f test-build.yml up \
          --abort-on-container-exit
</code></pre>
<p>As far as I can see, the envars are still not accessible from the containerized application.</p>
<p>Perhaps I am missing the obvious, but how would you go about passing secrets to a Compose build?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="why-not-try-calmer" data-post="3" data-topic="233474">
<div class="title">
<div class="quote-controls"></div>
 why-not-try-calmer:</div>
<blockquote>
<p>I’ve tested it on my project.</p>
</blockquote>
</aside>
<p>Specifically <em>how</em>? By creating a PR? Pushing to <code>master</code>? That’s what my question was about.</p>

**Comment by why-not-try-calmer** [comment]
<p>By pushing a commit to master.</p>

**Comment by airtower-luna** [comment]
<p>Okay, in that case it shouldn’t be the PR issue (but keep that in mind in case you want to accept PRs from forks).</p>
<p>Are the <code>environment</code> values used during the Docker Compose <code>build</code> stage? If yes, see this note in the <a href="https://docs.docker.com/compose/compose-file/compose-file-v3/#environment">Compose File docs</a>:</p>
<blockquote>
<p>If your service specifies a <a href="https://docs.docker.com/compose/compose-file/compose-file-v3/#build">build</a> option, variables defined in <code>environment</code> are <em>not</em> automatically visible during the build. Use the <a href="https://docs.docker.com/compose/compose-file/compose-file-v3/#args">args</a> sub-option of <code>build</code> to define build-time environment variables.</p>
</blockquote>
<p>There’s no other <em>obvious</em> problem at least. :sweat_smile:</p>

**Comment by why-not-try-calmer** [comment]
<p>Thanks for the suggestion, but I must admit I am a bit confused. When I want to build locally, I run <code>docker-compose</code> from my project’s root directory. It uses <code>docker-compose.yml</code> which does define:</p>
<pre><code class="lang-auto">build: .
environment:
  TELEGRAM_TOKEN: ${TELEGRAM_TOKEN}
  ...      
</code></pre>
<p>This works flawlessly: my program can access all the env vars defined there (notice this works because I have an .env file in that directory on which Compose can piggy-back to assign the appropriate environment variables passed to the build). So apparently I don’t need to use <code>args</code> in this situation. So by parity of reasoning, why would I need to use <code>args</code> in the situation where Compose is driven by a GitHub action? In both cases, the build is built in the same way.</p>
<p>Or perhaps the suggestion is to use <code>args</code> in the Github action? I don’t know how to, as the <code>args</code> property is not allowed under <code>steps</code>.</p>
<p>Anyway, I must be missing the obvious here because what I am trying to do is quite trivial. I guess I need to look at an example of someone exposing GitHub secrets to a Composed application driven by a GitHub action.</p>

**Comment by airtower-luna** [comment]
<p>I just tried passing a secret via environment myself, and it just worked:</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/airtower-luna/hello-github-actions/blob/6e319e3f8d50694e4ba1299b275eb588ed05766a/.github/workflows/secret.yaml#L36-L46">
  <header class="source">

      <a href="https://github.com/airtower-luna/hello-github-actions/blob/6e319e3f8d50694e4ba1299b275eb588ed05766a/.github/workflows/secret.yaml#L36-L46" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/airtower-luna/hello-github-actions/blob/6e319e3f8d50694e4ba1299b275eb588ed05766a/.github/workflows/secret.yaml#L36-L46" target="_blank" rel="noopener">airtower-luna/hello-github-actions/blob/6e319e3f8d50694e4ba1299b275eb588ed05766a/.github/workflows/secret.yaml#L36-L46</a></h4>



    <pre class="onebox">      <code class="lang-yaml">
        <ol class="start lines" start="36" style="counter-reset: li-counter 35 ;">
            <li>compose-secret:</li>
            <li>  runs-on: ubuntu-latest</li>
            <li>  needs: check-secret</li>
            <li>  if: needs.check-secret.outputs.secret_present == 'true'</li>
            <li>  steps:</li>
            <li>    - uses: actions/checkout@v3</li>
            <li>    - run: |</li>
            <li>        docker-compose up</li>
            <li>      working-directory: compose</li>
            <li>      env:</li>
            <li>        MY_ENV: '${{ secrets.TEST_SECRET }}'</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<aside class="quote no-group" data-username="why-not-try-calmer" data-post="7" data-topic="233474">
<div class="title">
<div class="quote-controls"></div>
 why-not-try-calmer:</div>
<blockquote>
<p>Or perhaps the suggestion is to use <code>args</code> in the Github action? I don’t know how to, as the <code>args</code> property is not allowed under <code>steps</code> .</p>
</blockquote>
</aside>
<p>No, I was wondering if the environment variables are supposed to be used in the <em>build</em>, in which case you’d need to use <code>build.args</code> in the compose file. But if that’s not the case the problem probably isn’t that. :slightly_smiling_face:</p>
<p>Combining my experiment and that information I think the most probable explanation is that for some reason the secrets aren’t actually available to the workflow. Naturally I can’t see your secrets, so I can’t actually verify that.</p>

**Comment by why-not-try-calmer** [comment, marked as answer]
<p>I was trying to invoke (GitHub’s jargon mode: ON) <em>environment</em> secrets when in fact I needed to invoke <em>repository secrets</em>. Doh. Sorry for wasting your time, and have a good one.</p>

**Reply by AnthonyDF** [reply]
Hey @why-not-try-calmer , could you please detail the modification you've done? I have the same issue and I don't understand your explanation: _I was trying to invoke (GitHub’s jargon mode: ON) environment secrets when in fact I needed to invoke repository secrets._
Thanks,

**Reply by OlaAlsaker** [reply]
I ran into the same issue, so I will give a little bit more context.

There is a difference between **environment secrets** and **repository secrets**. They are both accessible through the **Secrets and variables**-page in your repo's Settings.

## Environment secrets
To make these secrets available in your GitHub action, you have to specify which environment to use. These environments can be managed in the **Environments**-page in your repo's settings.

```yml
jobs:
  myJob:
    runs-on: ubuntu-latest
    environment: test
```

This action will now get access to all the secrets from my environment `test`, which then can be used like this:

```
${{ secrets.MY_SECRET }}
```

## Repository secrets
These secrets are **globally available for all your actions**, without having to specify any environment.

```
${{ secrets.MY_SECRET }}
```

So, what I _wanted_ was the repository secrets, but apparently I added my secrets to the _environment secrets_. Because of this, my action would fail, because my secret did not exist.

## TLDR
Make sure you use **repository secrets**, and not the **environment secrets**.
